File: online_judge/leetcode/n-queens.cc
#include<vector>
#include<string>

using namespace std;

class Solution {
public:
    vector<vector<string>> solveNQueens(int n) {
        vector<vector<string>> res;
        // initialize
        for (int i = 0; i < n; ++i) {
            vector<string> vs = vector<string>(n, string(n, '.'));
        }
        return res;
    }

    bool canPlace(vector<string> &vs, int x, int y) {
        int n = vs.size();
        for (int i = 0; i < n; ++i) {
            // horizontal or vertical
            if ((vs[x][i] == 'Q' && i != y) || (vs[i][y] == 'Q' && i != x)) {
                return false;
            }

            // anti diagonal
            if (x+i < n && y+i < n && vs[x+i][y+i] == 'Q') return false;
            if (x-i >= 0 && y-i >= 0 && vs[x-i][y-i] == 'Q') return false;

            // diagonal
            if (x-i >= 0 && y+i < n && vs[x-i][y+i] == 'Q') return false;
            if (x+i < n && y-i >=0 && vs[x+i][y-i] == 'Q') return false;
        }
        return true;
    }

    void dfs() {}
};

File: online_judge/leetcode/n-queens.py
class Solution(object):

    def __init__(self):
        self.res = []

    def solveNQueens(self, n):
        """
        :type n: int
        :rtype: List[List[str]]
        """
        board = ['.' * n for _ in range(n)]
        self.dfs(board, n, 0)
        return self.res

    def is_safe(self, board, x, y):
        n = len(board)
        for i in range(1, n):
            if x - i >= 0 and y + i < n:
                if board[x-i][y+i] == 'Q':
                    return False
            if x - i >= 0 and y - i >= 0:
                if board[x-i][y-i] == 'Q':
                    return False
            if x - i >= 0:
                if board[x-i][y] == 'Q':
                    return False
        return True

    def dfs(self, board, n, x):

        if x == n:
            self.res.append(list(board))
            return

        for i in range(n):
            if self.is_safe(board, x, i):
                board[x] = board[x][:i] + 'Q' + board[x][i+1:]
                self.dfs(board, n, x+1)
                board[x] = board[x][:i] + '.' + board[x][i+1:]
